Pulse drawer v2, mpl plotter and interface

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This adds a part of new pulse drawer (5/5). See #4643 
This PR implements one of plotters and common interface of the pulse drawer v2.


### Details and comments
The detail and the design concept are written in the module docstring.

- `Mpl2DPlotter` class is added. This is one of the plotter APIs of the v2 pulse drawer. This takes a canvas object and generate a 2D image from them.
- Several minor updates to generators and layouts to cleanup the appearance of output image.

You can test image generation with following code with this branch:
```python
from qiskit import QuantumCircuit, transpile, schedule
from qiskit.visualization.pulse_v2 import interface, stylesheet
from qiskit.test.mock import FakeAlmaden

qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)
qc.measure_all()
qc = transpile(qc, FakeAlmaden())
sched = schedule(qc, FakeAlmaden())

interface.draw(program=sched, style=stylesheet.IqxStandard(), backend=FakeAlmaden())
```

You will get this image with above code.

![image](https://user-images.githubusercontent.com/39517270/93883530-a8bb9a00-fd1c-11ea-992e-8c738d57c7f2.png)

This is much cleaner than the existing drawer's output.
![image](https://user-images.githubusercontent.com/39517270/93883638-ce48a380-fd1c-11ea-9a87-5c288ddd8adc.png)

In addition, the wall time to generate above image is:
- v2 drawer: 143 ms
- v1 drawer: 660 ms

This performance improvement is thanks to the waveform truncation and consecutive data points compression.

We need to write reno after sufficient usability test until official release (I don't think I did enough test with this drawer, because it has lots of new features), add import to the top level visualization module (also add this to API docs), and support parametrized schedule visualization with #5073 .

lcapelluto commented on qiskit/visualization/pulse_v2/interface.py:
```diff
@@ -0,0 +1,415 @@
+         show_framechange: bool = True,
+         show_waveform_info: bool = True,
+         show_barrier: bool = True,
+         plotter: str = types.Plotter.Mpl2D,
```
Is this more accurate?
```suggestion
         plotter: Union[str, types.Plotter] = types.Plotter.Mpl2D,
```

lcapelluto replied:
or should it be
```suggestion
         plotter: str = types.Plotter.Mpl2D.value(),
```
?

lcapelluto commented on qiskit/visualization/pulse_v2/drawing_objects.py:
```diff
@@ -70,7 +71,7 @@ class ElementaryData(ABC):
 
     def __init__(self,
-                 data_type: str,
+                 data_type: Union[str, Enum],
```
Are these any Enum, or are they one of the enums in `types.py`? I think you could use, for example, `types.DynamicString` as the type

nkanazawa1989 replied:
I assume end-user sets a custom data type. `str` is the simplest way to define custom type. basically this can be anything that is hashable.

lcapelluto replied:
You're specifically adding `Enum`, though. Why add it?

lcapelluto commented:
When can we replace `schedule.draw` with this?

nkanazawa1989 commented on qiskit/visualization/pulse_v2/generators/waveform.py:
```diff
@@ -294,39 +296,37 @@ def gen_waveform_max_value(data: types.PulseInstruction,
     # max of real part
     re_maxind = np.argmax(np.abs(ydata.real))
-    if np.abs(ydata.real[re_maxind]) > formatter['general.vertical_resolution']:
+    if np.abs(ydata.real[re_maxind]) > 0.01:
```
because this generator shows only 2 digits after the decimal point.

lcapelluto replied:
Maybe add a comment?

nkanazawa1989 commented on qiskit/visualization/pulse_v2/stylesheet.py:
```diff
@@ -196,12 +97,14 @@ class IqxStandard(dict):
         super().__init__()
         style = {'formatter.control.apply_phase_modulation': True,
                  'formatter.control.show_snapshot_channel': True,
+                 'formatter.control.show_acquire_channel': False,
```
I want to feedback from users about this. I always remove them since it doesn't have much information except for some edge cases, for example, tuning of acquisition timing in the context of calibration. but if you want to check them this can be True by default.

lcapelluto replied:
I'm not sure. @taalexander where do you weigh in on this?

lcapelluto replied:
We will leave default to True, but certain style sheets should probably set to false. We can open a `discussion` issue on github to get feedback and ideas from users

nkanazawa1989 commented:
@lcapelluto thanks for reviewing.

Now the drawer supports visualization of parametric pulse with unbound parameters. I also fixed a bug of sequential truncation along the same channel and updated unittests.

This is example of generated image:

```python
from qiskit import QuantumCircuit, transpile, schedule, pulse
from qiskit.circuit import Parameter

from qiskit.visualization.pulse_v2 import interface, stylesheet
from qiskit.test.mock import FakeAlmaden


with pulse.build(FakeAlmaden()) as cx10:
    pulse.play(pulse.GaussianSquare(3200, Parameter('amp_cx'), 128, 2688), pulse.DriveChannel(0))
    pulse.play(pulse.GaussianSquare(3200, 0.1, Parameter('sigma_cx'), 2688), pulse.ControlChannel(1))

qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)
qc.measure_all()
qc.barrier(0, 1)
qc.cx(1, 0)
qc.barrier(0, 1)
qc.measure_all()
qc.add_calibration('cx', (1, 0), cx10)

qc = transpile(qc, FakeAlmaden())
sched = schedule(qc, FakeAlmaden())

interface.draw(program=sched, style=stylesheet.IQXStandard(), backend=FakeAlmaden())
```

![image](https://user-images.githubusercontent.com/39517270/96299350-e4513780-102e-11eb-81db-373ddb588ed1.png)

I think this PR is ready for the final review.

taalexander commented on qiskit/visualization/pulse_v2/layouts.py:
```diff
@@ -40,47 +40,44 @@ def my_channel_layout(channels: List[pulse.channels.Channel],
+3. layout.figure_title
+
+An end-user can write arbitrary functions that output the string data that
+will be later consumed by the plotter API to output figure title.
```
```suggestion
will be later consumed by the plotter API to output the figure title.
```

taalexander commented on qiskit/visualization/pulse_v2/layouts.py:
```diff
@@ -40,47 +40,44 @@ def my_channel_layout(channels: List[pulse.channels.Channel],
+
+An end-user can write arbitrary functions that output the string data that
+will be later consumed by the plotter API to output figure title.
+Layout function in this module are called with the `program` and `device` kwargs.
```
```suggestion
Layout functions in this module are called with the `program` and `device` kwargs.
```

taalexander commented on qiskit/visualization/pulse_v2/layouts.py:
```diff
@@ -40,47 +40,44 @@ def my_channel_layout(channels: List[pulse.channels.Channel],
+An end-user can write arbitrary functions that output the string data that
+will be later consumed by the plotter API to output figure title.
+Layout function in this module are called with the `program` and `device` kwargs.
+These data provides input program and backend system configuration.
```
```suggestion
This data provides input program and backend system configurations.
```

taalexander commented on qiskit/visualization/pulse_v2/layouts.py:
````diff
@@ -40,47 +40,44 @@ def my_channel_layout(channels: List[pulse.channels.Channel],
     ```
 
-The user can set the custom stylesheet to the drawer interface.
+Arbitrary layout function satisfying the above format can be accepted.
````
```suggestion
An arbitrary layout function satisfying the above format can be accepted.
```

taalexander commented on qiskit/visualization/pulse_v2/layouts.py:
```diff
@@ -355,8 +356,42 @@ def time_map_in_ns(time_window: Tuple[int, int],
+
+def empty_title(program: Union[pulse.Waveform, pulse.ParametricPulse, pulse.Schedule],
+                device: DrawerBackendInfo) -> str:
+    """Layout function for generating figure title.
+
+    Returns empty title.
+    """
```
```suggestion
    """Layout function for generating an empty figure title."""
```

taalexander commented on qiskit/visualization/pulse_v2/plotters/matplotlib.py:
```diff
@@ -0,0 +1,201 @@
+        self.ax.set_yticklabels([])
+        self.ax.yaxis.set_tick_params(left=False)
+
+    def draw(self):
```
What will happen if this method is called twice?

nkanazawa1989 replied:
Plotter is initialized for each drawer call and thus draw is called only once. For dynamic image update we can use another plotter. If we call this twice with some hack perhaps all image components are drawn twice because I don't clear figure.

taalexander commented on qiskit/visualization/pulse_v2/plotters/matplotlib.py:
```diff
@@ -0,0 +1,201 @@
+        Args:
+            filename: File path to output image data.
+        """
+        plt.savefig(filename,
```
This is based on the current global matplotlib figure, rather this Class should own a figure that it saves.

nkanazawa1989 replied:
right, can this savefig method be removed (offloaded to user) or I need to fix this and keep the method?

taalexander replied:
I think its good to have this as a method, I just think it should be called with https://matplotlib.org/api/_as_gen/matplotlib.figure.Figure.html#matplotlib.figure.Figure.savefig

taalexander commented on test/python/visualization/pulse_v2/test_events.py:
```diff
@@ -98,3 +98,47 @@ def test_frequency(self):
         self.assertAlmostEqual(inst_data1.frame.freq, 1.0)
+
+    def test_parametrized_parametric_pulse(self):
+        """Test generating waveforms that is parameterized."""
```
```suggestion
        """Test generating waveforms that are parameterized."""
```

taalexander commented on test/python/visualization/pulse_v2/test_drawings.py:
```diff
@@ -0,0 +1,185 @@
+        self.assertEqual(data1, data2)
+
+    def test_text_data_with_abstract_coordinate(self):
+        """Test for TextData with abstract coordinate."""
```
```suggestion
        """Test for TextData with abstract coordinates."""
```

taalexander commented on test/python/visualization/pulse_v2/test_events.py:
```diff
@@ -98,3 +98,47 @@ def test_frequency(self):
+        self.assertEqual(pulse_inst.inst, test_waveform)
+
+    def test_parametrized_frame_change(self):
+        """Test generating waveforms that is parameterized.
```
```suggestion
        """Test generating waveforms that are parameterized.
```

taalexander commented on test/python/visualization/pulse_v2/test_events.py:
```diff
@@ -98,3 +98,47 @@ def test_frequency(self):
+
+        Parametrized phase should be ignored when calculating waveform frame.
+        This is due to phase modulated representation of waveforms,
+        i.e. we cannot calculate phase factor of waveform if phase is unbound.
```
```suggestion
        i.e. we cannot calculate the phase factor of waveform if the phase is unbound.
```

taalexander reviewed (changes requested):
Looks excellent, just some minor comments and suggestions.

nkanazawa1989 commented on qiskit/visualization/pulse_v2/generators/frame.py:
```diff
@@ -234,32 +233,40 @@ def gen_raw_operand_values_compact(data: types.PulseInstruction,
+            base=data.frame.freq / (10**int(np.floor(np.log10(data.frame.freq)))),
+            exp=int(np.floor(np.log10(data.frame.freq)))
+        )
+    frame_info = '{phase:.2f}\n{freq}'.format(phase=data.frame.phase,
```
It's intentional because of limited horizontal space (it overlaps with other FC labels and cannot read).

nkanazawa1989 commented on qiskit/visualization/pulse_v2/interface.py:
```diff
@@ -0,0 +1,447 @@
+        axis: Arbitrary object passed to the plotter. If this object is provided,
+            the plotters uses given `axis` instead of internally initializing a figure object.
+            This object format depends on the plotter. See plotters section for details.
+        filename: If provided the output image is dumped into a file under the filename.
```
You mean the drawer just need to generate an image and the file save should be done on user side?

taalexander replied:
Yes 👍

taalexander commented on qiskit/visualization/pulse_v2/types.py:
```diff
@@ -212,6 +212,16 @@ class Plotter(str, Enum):
+    SYSTEM_CYCLE_TIME: System time dt.
+    NANO_SEC: Nano seconds.
+    """
+    SYSTEM_CYCLE_TIME = 'dt'
+    NANO_SEC = 'ns'
```
Maybe `CYCLES` and `NS` or is that less clear?

taalexander commented:
PR appears to be failing CI. Once its passing we can get it in 👍 .

nkanazawa1989 commented:
Thanks. I just removed file save method because the file save method is called only by interface (i.e. top level `draw`, because `draw` doesn't return plotter class) and managing file save API arguments of different plotters with `draw` arguments may be difficult (or make arguments messy). The function returns a figure object and user can save it in file if the figure object has a method to save itself (matplotlib Figure has `.savefig`).
